harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+        check.is_not_none(self.model, "Must have one model")
+        check.gt_eq(len(self.optimizers), 1, "Must have at least one optimizer")
+
+    def _Model(self, model: torch.nn.Module) -> torch.nn.Module:
+        self.model = model.to(self.device)
+        if not self.hvd_config.use and self.n_gpus > 1:
+            check.eq(
+                self.hvd_config.aggregation_frequency,
+                1,
+                "Please enable `optimized_parallel` to use aggregation "
+                "frequency greater than 1 for single machine multi-GPU "
+                "training.",
+            )
+            self.model = nn.DataParallel(self.model)
+            logging.debug("Initialized mode for native parallel training.")
+        return model
+
+    def _Optimizer(self, optimizer: torch.optim.Optimizer) -> torch.optim.Optimizer:  # type: ignore
+        check.is_not_none(self.model, "Must have one model first.")
+
+        # TODO: Check that optimizer is not an amp optimizer.
+        if self.hvd_config.use:
+
+            def filter_named_parameters(
+                model: nn.Module, optimizer: torch.optim.Optimizer,  # type: ignore
+            ) -> List:
+                """
+                filter_named_parameters filters the named parameters of a specified optimizer out
+                of all the named parameters from a specified model. We need this function because
+                a torch.optim.Optimizer doesn't store parameter names and we need the names of
+                the parameters when mapping parameters to each horovod.DistributedOptimizer.
+                """
+                opt_params = {
+                    p for group in optimizer.param_groups for p in group.get("params", [])
+                }
+                return [(name, p) for name, p in model.named_parameters() if p in opt_params]
+
+            use_compression = self.hvd_config.fp16_compression
+            optimizer = hvd.DistributedOptimizer(
+                optimizer,
+                named_parameters=filter_named_parameters(cast(nn.Module, self.model), optimizer),
+                backward_passes_per_step=self.hvd_config.aggregation_frequency,
+                compression=hvd.Compression.fp16 if use_compression else hvd.Compression.none,
+            )
+            logging.debug("Initialized optimizer for distributed and optimized parallel training.")
+
+        self.optimizers.append(optimizer)
+        return optimizer
+
+    def _LRScheduler(
+        self,
+        lr_scheduler: torch.optim.lr_scheduler._LRScheduler,
+        step_mode: pytorch.LRScheduler.StepMode,
+    ) -> pytorch.LRScheduler:
+        # Initialize LR schedulers. This block must happen after wrapping the optimizers using
+        # horovod.DistributedOptimizer because torch.optim._LRScheduler directly access the
+        # optimizer parameter groups, which are also Horovod.DistributedOptimizer relies on.
+        wrapped = pytorch.LRScheduler(lr_scheduler, step_mode)
```

Comment:
blocking: we should perform a check there that the optimizer that the scheduler is pointing at is one of the optimizers we returned to the user.